Add tests for dsname in datasource

ad-m opened the pull request:
I have a number of small changes. Please review commit-by-commit, since the appropriate separation has been preserved.

If combined commits is a problem - it will create several separate PRs.

ad-m commented on cloudinit/apport.py:
```diff
@@ -34,12 +34,14 @@
     'OpenStack',
     'Oracle',
     'OVF',
+    'RbxCloud (HyperOne, Rootbox, Rubikon)',
```
Added RbxCloud according guides how to create datasource.

OddBloke replied:
There are a couple of other entries here that have similar formatting:

```
    'GCE - Google Compute Engine',
    'IBM - (aka SoftLayer or BlueMix)',
```

Should this perhaps be 
```suggestion
    'RbxCloud - (HyperOne, Rootbox, Rubikon)',
```

for consistency?

ad-m commented on cloudinit/sources/DataSourceRbxCloud.py:
```diff
@@ -153,6 +153,7 @@ def read_user_data_callback(mount_dir):
 
     data = {
         'userdata': user_data,
+        'vendordata': meta_data,
```
Added to expose vendordata to fullfil datasource API.

OddBloke replied:
vendordata is cloud-provided cloud-config (per https://cloudinit.readthedocs.io/en/latest/topics/vendordata.html). Looking at the way that `meta_data` is used in this function, it looks like it's a dictionary loaded from JSON, which isn't the correct format for vendordata.

Am I missing something?

ad-m commented on cloudinit/sources/DataSourceRbxCloud.py:
```diff
@@ -197,6 +198,7 @@ def read_user_data_callback(mount_dir):
 
 
 class DataSourceRbxCloud(sources.DataSource):
+    dsname = "RbxCloud"
```
Fix bug https://bugs.launchpad.net/cloud-init/+bug/1855196 by add dsname.

ad-m replied:
Should I mark my commit in special way to reference in launchpad?

OddBloke replied:
We'll want `LP: #1855196` in the commit message that ends up in-tree.

ad-m commented on cloudinit/sources/DataSourceRbxCloud.py:
```diff
@@ -217,6 +219,7 @@ def _get_data(self):
         """
         rbx_data = get_md()
         self.userdata_raw = rbx_data['userdata']
+        self.vendordata_raw = rbx_data['vendordata']
```
Expose vendordata to fullfill datasource API.

ad-m commented on doc/rtd/topics/datasources.rst:
```diff
@@ -30,12 +30,12 @@ The following is a list of documents for each supported datasource:
    datasources/configdrive.rst
    datasources/digitalocean.rst
+   datasources/ec2.rst
+   datasources/e24cloud.rst
```
Sorted alphabetically.

OddBloke replied:
These two, specifically, are still in reverse alphabetical order, I think.

ad-m replied:
It depends whether the numbers come before or after the letters. The alphabet does not define it, so I will drop that change.

OddBloke replied:
For clarity, this is my basis for saying they are in reverse order:

```shell
$ sort << EOF
>    datasources/ec2.rst
>    datasources/e24cloud.rst
> EOF
   datasources/e24cloud.rst
   datasources/ec2.rst
```

(Or do you get a different order with a different locale?)

ad-m commented on doc/rtd/topics/datasources.rst:
```diff
@@ -45,6 +45,7 @@ The following is a list of documents for each supported datasource:
    datasources/openstack.rst
    datasources/oracle.rst
    datasources/ovf.rst
+   datasources/rbxcloud.rst
```
Added missing.

ad-m commented on doc/rtd/topics/datasources/rbxcloud.rst:
```diff
@@ -1,4 +1,4 @@
-.. _datasource_config_drive:
+.. _datasource_rbx:
```
Fix anchor (compare with filename).

ad-m commented on tests/unittests/test_datasource/test_common.py:
```diff
@@ -95,5 +97,11 @@ def test_data_sources_have_valid_network_config_sources(self):
 
+    def test_expected_dsname_defined(self):
+        for ds in DEFAULT_LOCAL + DEFAULT_NETWORK:
+            fail_msg = ('{} has an invalid / missing dsname property'
```
Avoid bug like https://bugs.launchpad.net/cloud-init/+bug/1855196 .

OddBloke commented on tests/unittests/test_datasource/test_common.py:
```diff
@@ -95,5 +97,12 @@ def test_data_sources_have_valid_network_config_sources(self):
+                '{} has an invalid / missing dsname property'.format(str(ds))
+            )
+            self.assertNotEqual(ds.dsname, DataSource.dsname, fail_msg)
+            self.assertNotEqual(ds.dsname, None)
```
```suggestion
            self.assertIsNotNone(ds.dsname)
```

OddBloke commented on tests/unittests/test_datasource/test_common.py:
```diff
@@ -95,5 +97,12 @@ def test_data_sources_have_valid_network_config_sources(self):
+    def test_expected_dsname_defined(self):
+        for ds in DEFAULT_LOCAL + DEFAULT_NETWORK:
+            fail_msg = (
+                '{} has an invalid / missing dsname property'.format(str(ds))
```
If we're going to override the default message, I think we should include at least `ds.dsname` in the message.

OddBloke commented on cloudinit/settings.py:
```diff
@@ -26,6 +26,7 @@
         'Azure',
         'AltCloud',
         'OVF',
+        'RbxCloud',
```
This should be added at the end, so that people somehow relying on the existing order don't see their behaviour modified.

ad-m commented:
@OddBloke , I do not see a problem to properly merge commity after applying patches.

ad-m commented:
> I don't fully understand. We don't do merge commits in cloud-init's master. Are you saying you're happy with a squash merge?

I modified the previous commits to include the expected changes (using rebase & fixup).

ad-m commented:
> Dropping the vendor-data change was intended, I assume?
Yes.
> I have some review comments on them too.
Fixed.
> that's a nit and I don't think it's worth the effort of picking them apart again.
Splited.
> This doesn't look resolved to me.
Fixed.

OddBloke commented on tests/unittests/test_datasource/test_common.py:
```diff
@@ -95,5 +97,14 @@ def test_data_sources_have_valid_network_config_sources(self):
+                )
+            )
+            self.assertNotEqual(ds.dsname, DataSource.dsname, fail_msg)
+            self.assertIsNotNone(ds.dsname, None)
```
This still isn't quite right:

```suggestion
            self.assertIsNotNone(ds.dsname)
```

ad-m commented:
@OddBloke, i created seperate PR to fix that in #118 .

OddBloke commented:
@ad-m This requires rebasing on master.

(In retrospect, doing this all in one PR is something of a pain, we should do separate PRs in future. Still, this was a valuable exercise to work that out!)

ad-m commented:
@OddBloke, fixed.
